Commands for generating shell completion scripts + statically-generated bash script

baronfel opened the pull request:
Part of #42397 




Implements:

* provider model for shell completions
* new 'completions' command with 'script' and 'register' subcommands

## Shell Support

| shell | recording | static completions | dynamic completions (e.g. nuget packages) | slower, fallback completions |
| - | - | - | - | - |
| bash | [🎥](https://asciinema.org/a/696566) | ✔️ | ✔️ | ❌  |
| zsh | [🎥](https://asciinema.org/a/696567)|  ✔️ | ✔️ | ❌ |
| pwsh | [🎥](https://asciinema.org/a/695838) | ✔️ | ✔️ | ❌ |
| nushell | ❌ | ❌ | ❌ | ✔️ 
| fish | ❌ | ❌ | ❌ | ✔️ |


* for nushell/fish, the generate command just generates a correct-but-slow fallback wrapper
* for bash, zsh, and powershell, comprehensive and very fast static completions have been implemented, even for dynamic completions like nuget packages
* for zsh and powershell, _descriptions_ of completion items are shown in nice menus

baronfel committed:
Spike out initial completion script emit commands

baronfel committed:
Update contents to include locations and commented instructions where possible

baronfel committed:
Complete rewrite to support dynamic completions

baronfel commented:
I've implemented completion script generate for bash based partially on the _excellent_ patterns that the [Swift ArgumentParser library](https://github.com/apple/swift-argument-parser/blob/main/Sources/ArgumentParser/Completions/BashCompletionsGenerator.swift#L148) has used. Here's an example of the generated script (warning, 3500 lines of bash await you):

[dotnet-bash-completions.txt](https://github.com/user-attachments/files/16404032/dotnet-bash-completions.txt)

It works _remarkably_ well - one of the harder parts was figuring out when truly dynamic completions were required and injecting a call to the dotnet CLI's completion command to generate suggestions for those cases. But where completions can be known statically we inject them.  There is no good handling for file-based completions yet - one hurdle I faced was the very-pluggable completions model currently doesn't give any sense of attribution of the different completion sources - if I knew a given `Func<CompletionContext, CompletionItem[]>` was a certain kind of completion source I could maybe skip them for purposes of static script generation. As it is I had to create my own signal for that kind of data.

baronfel committed:
Refactor to make certain scary portions of the generation have nicer names.

Addresses the final addressable warning about top-level matching on -?.
-? is a pattern in bash, so it needs to be escaped.

baronfel committed:
Enhancement: do not emit completions for option arguments for flag-style options

baronfel committed:
Optimize static completions for bash: don't emit option completions for options without any completions.

baronfel committed:
Completion optimization - don't emit option handler stanzas if they are empty

baronfel committed:
Completion optimization: mark several flag-style commands as Arity zero to prevent them from being generated.

This is breaking from the parsing perspective, but all of these have single-digit usages on GitHub.com

baronfel committed:
Completions optimization: mark many flag-style options as Arity Zero

Technically breaking from a parser perspective, but GitHub searches of usages for these all showed either zero with arguments or ~5.

baronfel committed:
Fix accidental zero-arity arguments in tool install and restore

baronfel committed:
update baselines for new completions command

baronfel committed:
strip odd help option aliases, suggest only best option name, and ensure we include recursive options from parents

baronfel committed:
Add tests for shell completion helpers

baronfel commented on src/Cli/dotnet/commands/dotnet-completions/shells/Bash.cs:
```diff
@@ -0,0 +1,322 @@
+            writer.Indent++;
+            foreach (var subcommand in visibleSubcommands)
+            {
+                writer.WriteLine($"({subcommand.Name})");
```
I'm not 100%-solid on my bash-isms for what the difference is between this form and the form below - in this case I was translating some of the forms from the Swift static completions I linked above. For example, [here](https://github.com/apple/swift-argument-parser/blob/main/Sources/ArgumentParser/Completions/BashCompletionsGenerator.swift#L100) is the subcommand case generation and [here](https://github.com/apple/swift-argument-parser/blob/main/Sources/ArgumentParser/Completions/BashCompletionsGenerator.swift#L165) is the option case generation.

baronfel replied:
Thanks for the context - I'm quite happy to switch to _`(pattern)`_ everywhere. One thing I have to do on this PR is add acceptance tests to pin down the generation, so if we want to make correctness/stylistic changes in the future we'll be able to track them more easily.

baronfel committed:
WIP test placeholders

baronfel committed:
Rename for consistency

baronfel committed:
Initial powershell static completions provider

This provider handles option and subcommand name completions,
as well as argument completions with known values, but doesn't
handle
 * options or arguments that have dynamic values
 * triggering argument/option completions aside from the name

baronfel committed:
Powershell: Generate dynamic completion lists for arguments where possible.

baronfel committed:
Extract out implementation to separate project to potentially ship as a standalone library

Also extract out the tests to make it easier to iterate.

baronfel committed:
Initial basic snapshot tests for bash and pwsh

baronfel committed:
Minor test framework refactor

baronfel commented on src/System.CommandLine.StaticCompletions/shells/ZshShellProvider.cs:
```diff
@@ -1,34 +1,302 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.CodeDom.Compiler;
+using Microsoft.VisualBasic;
```
Unnecessarily added by the IDE on an errand type name completion - I use 'let' a lot due to my F# mindset and that auto completes in C# to a type from this namespace. It'll be cleaned up before this is up for review properly.

baronfel committed:
Working zsh completions (including testing on dotnet)

baronfel committed:
ZSH: fix completions for static argument completions with descriptions

baronfel committed:
Zsh: support for dynamic completions of options and arguments

baronfel commented:
Ok, I've updated the description. This is now ready for a review + discussion.

I've done almost all of the work in a separate library project - this is something that we should consider spinning off in the future, as the only real dependency it has is System.CommandLine. If that becomes stable then it'd be fitting for a library.

Assuming the library is implemented (and there are some baseline tests added) then the usage of the library in the dotnet CLI becomes very simple.

Forgind commented on src/Cli/dotnet/CommonOptions.cs:
```diff
@@ -109,25 +111,22 @@ public static CliArgument<string> DefaultToCurrentDirectory(this CliArgument<str
             {
-                Description = CommonLocalizableStrings.DisableBuildServersOptionDescription
+                Description = CommonLocalizableStrings.DisableBuildServersOptionDescription,
+                Arity = ArgumentArity.Zero
```
Does adding Arity = ArgumentArity.Zero in all these places actually do anything? Prevent you from saying --disable-build-servers:true? Would we want to prevent that?

baronfel replied:
This is where things get interesting. You're right that we're technically preventing this syntax, but this is a forwarded option, one whose whole purpose is to be forwarded as MSBuild properties. This means we have to look at the forwarding function to see if the flag value even matters - and in this case the forwarding function is hard-coded to not look at the boolean's value at all. In this case, the currently-chosen semantic is a flag, not an option with an argument. Knowing this makes the completions generation better in some cases, so I've nailed down the behavior in some scenarios like this.

Forgind commented on src/Cli/dotnet/CommonOptions.cs:
```diff
@@ -277,6 +279,18 @@ internal static CliArgument<T> AddCompletions<T>(this CliArgument<T> argument, F
             return argument;
         }
+
+        internal static DynamicOption<T> AddCompletions<T>(this DynamicOption<T> option, Func<CompletionContext, IEnumerable<CompletionItem>> completionSource)
```
Can this take an IDynamicOption?

baronfel replied:
IMO no because then you'd lose the knowledge of the generic `T` for the DynamicOption. DynamicOption<T> is a .NET CLI concept, whereas IDynamicOption is a concept of the completions library I added.

Forgind commented on src/Cli/dotnet/commands/dotnet-add/AddCommandParser.cs:
```diff
@@ -13,7 +13,8 @@ internal static class AddCommandParser
         public static readonly CliArgument<string> ProjectArgument = new CliArgument<string>(CommonLocalizableStrings.ProjectArgumentName)
         {
-            Description = CommonLocalizableStrings.ProjectArgumentDescription
+            Description = CommonLocalizableStrings.ProjectArgumentDescription,
```
I think I do this in one of my PRs, so conflict, but that doesn't matter too much.

Forgind commented on src/Cli/dotnet/commands/dotnet-add/AddCommandParser.cs:
```diff
@@ -13,7 +13,8 @@ internal static class AddCommandParser
         {
-            Description = CommonLocalizableStrings.ProjectArgumentDescription
+            Description = CommonLocalizableStrings.ProjectArgumentDescription,
+            Hidden = true
```
Why no Arity for this? It seems like you added it to a lot of things.

baronfel replied:
You get some arities by default in System.CommandLine based on the type of the CliArgument/CliOption - this is one of those cases. You can see the logic used [here](https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/ArgumentArity.cs#L140C1-L166C6). In this case, I just needed to hide the argument to paper over some completions pain points. Since we think very few users use this arg (this is one of the discussion points in your verb/noun PR, remember!) this should be fine while also not breaking existing users.

baronfel replied:
Regarding the ArgumentArity.Zero stuff you see around this PR - We have a lot of _flags_ in the CLI. We don't currently have a unified way of modeling those (i.e. a CliOption<T> subclass). It might be nice to do that (and/or S.CL to have that concept already)

Forgind commented on src/Cli/dotnet/commands/dotnet-complete/CompleteCommandParser.cs:
```diff
@@ -14,6 +14,12 @@ internal static class CompleteCommandParser
+        public static readonly CliOption<bool> Detailed = new("--detailed")
+        {
+            Hidden = true,
+            Description = "Show detailed completion information"
```
Localize?

baronfel replied:
Good call 👍

baronfel replied:
Actually that new option was vestigial as well - once we have completion script generation hooked up for users we can

a) add a new output format for `dotnet complete` that includes descriptions for the returned items, if any
b) update the shell providers that understand dynamic completions (currently bash and zsh) to use the new flag so that they can provide the descriptions and not just the values. this would be a nice UX improvement.

Forgind commented on src/Cli/dotnet/completions.ps1:
```diff
@@ -0,0 +1,1383 @@
+    switch ($command) {
+        'dotnet' {
+            $staticCompletions = @(
+                [CompletionResult]::new('--help', '--help', [CompletionResultType]::ParameterName, "Show command line help.")
```
Are these auto-generated somehow? How are they maintained? Should the descriptions be localized?

baronfel replied:
I actually need to remove this, it shouldn't be checked in here. I do need to add snapshots for the .NET CLI grammar to prevent unexpected drift now that you mention it!

Forgind commented on src/System.CommandLine.StaticCompletions/CompletionsCommand.cs:
```diff
@@ -0,0 +1,105 @@
+        {
+            if (OperatingSystem.IsWindows())
+            {
+                return shellMap[PowershellShellProvider.PowerShell];
```
I don't normally use pwsh on windows...

baronfel replied:
I'm guessing you use `cmd`? First off, it's 2024 (almost 2025 dude 😛). Secondly, `cmd` has no good completions story anyway so there's no sense pretending it even exists for this feature. Finally, this is just the default value computation here - I'd expect users that worry about drift here to add an actual value for the shell a la `dotnet completions generate pwsh` in their $PROFILE or equivalent.

Forgind replied:
Hey, some people use vim...

I'm not really worried about the lack of completions; I just want to make sure that we don't make assumptions that end up causing issues for people using unexpected shells. I didn't test what would happen if I tried to use this on cmd, so if it just doesn't do anything, that's fine, but if it tries to show completions and explodes, that's not fine.

When I looked at your demos in the description, it seemed like you had to explicitly run the completions.ps1 script to get completions. Is that the only way to ever get completions? I kinda thought there was like a tab-complete option or something, and if so, I'd just want that to not break current cmd users.

Forgind commented on src/System.CommandLine.StaticCompletions/CompletionsCommand.cs:
```diff
@@ -0,0 +1,105 @@
+                throw new InvalidOperationException($"Shell '{shellPath}' is not supported");
+            }
+
+            return shellProvider;
```
Unreachable? Does C# really not realize that?

baronfel replied:
Good catch, dunno why my editor didn't realize that.

Forgind commented on src/System.CommandLine.StaticCompletions/CompletionsCommand.cs:
```diff
@@ -0,0 +1,105 @@
+            {
+                throw new InvalidOperationException("Could not determine the shell from the environment");
+            }
+            var shellName = Path.GetFileName(shellPath);
```
nit: blank space?

Forgind commented on src/System.CommandLine.StaticCompletions/CompletionsCommand.cs:
```diff
@@ -0,0 +1,105 @@
+    }
+}
+
+public class RegisterScriptCommand : CliCommand
```
Can you explain the value of this command? Seems like it does strictly less than GenerateScript

baronfel replied:
This is some ground work that I should probably delete because it's vestigial at this time. The thinking is we should have an explicit command that users run to generate the completion script for their shell, and that makes total sense to do. Users that want completions can open their Powershell $PROFILE and add `dotnet completions generate pwsh | Out-String | Invoke-Expression` to it and have up-to-date completions each time they load their shell. This can of course be optimized, etc. but this is the main idea of many shell-completion frameworks that exist today.

However, you still need to get users to _do_ that step. This is why some other frameworks have an 'install' and 'uninstall' semantic. This looks something like `dotnet completions register pwsh`, and the logic would be, roughly:

* determine the shell the user is requesting for
* ask the shell provider to see if the dotnet completions are registered
  * this will be heuristic-based, likely, and require probing contents of files like $PROFILE for pwsh, .zshrc for Zsh, etc.
* if not, inject a section to call the `dotnet completions generate` command into the shell configuration files
* (optionally) support uninstalling as well - do the same but find and remove the special section

The 'fuzziness' of this command is why I haven't implemented it yet. But once it exists, we could then have a discussion about if the SDK first-run experience should set up this completion registration automatically - we know today that so very very few of our users use completions, and having them set up would be a huge usability increase - as well as make it so that any fixes/enhancement to the completions we might make would be automatically applied!

Forgind replied:
I like the idea of a dotnet completions register pwsh command, but even more than that, I like the idea of making this part of the first-run welcome message but in a slightly different way. I'm envisioning two options:
1. A message like "If you want to see what options are available, run 'dotnet completions register pwsh'."
2. Optionally pause before executing their command with a question like "do you want completions? y/n" and let them choose.

I think something like that would dramatically increase usage, and I do agree with "huge usability increase" _if_ people start using it...but I also don't like the idea of opting people into something without telling them. I actually do see an argument in favor of not having completions, namely that it can take up a lot of screen real estate, and it may not ultimately be helpful if we do a poor job naming our subcommand/options, and people feel the need to go online to read more about them anyway. As we're busy telling people how to opt in (or giving them the y/n option), we can make sure to mention something like "if you ever don't want this anymore, run 'dotnet completions unregister pwsh'." Though I'll admit that unregister might go the way of unregister in MSBuildLocator in becoming vestigial, I think it's positive to offer it as an option initially. It makes it less scary to opt into something if I know I can change my mind later.

As a small aside, I like install/uninstall more than register/unregister, but maybe that's just me.

Forgind commented on src/System.CommandLine.StaticCompletions/DynamicArgument.cs:
```diff
@@ -0,0 +1,10 @@
+/// Represents an Argument whose completions are dynamically generated and so should not be emitted in static completion scripts.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public interface IDynamicArgument;
```
Should this class (and the next) be IDynamicArgument.cs and IDynamicOption.cs?

Forgind commented on src/System.CommandLine.StaticCompletions/HelpGenerationExtensions.cs:
```diff
@@ -0,0 +1,120 @@
+    };
+
+    /// <summary>
+    /// sanitizes a function name to be safe for bash
```
tiny nit:
```suggestion
    /// Sanitizes a function name to be safe for bash
```

Forgind commented on src/System.CommandLine.StaticCompletions/HelpGenerationExtensions.cs:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    public static IEnumerable<CliOption> HeirarchicalOptions(this CliCommand c)
```
nit:
```suggestion
    public static IEnumerable<CliOption> HierarchicalOptions(this CliCommand c)
```

Forgind commented on src/System.CommandLine.StaticCompletions/System.CommandLine.StaticCompletions.csproj:
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(SdkTargetFramework)</TargetFramework>
```
I think this would have to change if you wanted to make it a separate library

baronfel replied:
Totally agree - this was one of my few compromises in this regard because I wasn't sure how long it might be before we could release this. We'd need a stable S.CL version before we could reasonably do that, and I don't know a timeline for that at the moment.

baronfel replied:
Totally agree - this was one of my few compromises in this regard because I wasn't sure how long it might be before we could release this. We'd need a stable S.CL version before we could reasonably do that, and I don't know a timeline for that at the moment.

Forgind reviewed (approved):
I didn't look at shell-specific code or test code, but this looks great! I appreciated the videos in the description, but you might need to work on your typing speed (or crop the first half of the video) 😉

baronfel committed:
Code review fixes and renames from @Forgind

baronfel committed:
Remove unused option to minimize diff

baronfel commented on src/Cli/dotnet/Parser.cs:
```diff
@@ -53,18 +53,31 @@ public static class Parser
+        public static readonly CliOption<bool> VersionOption = new("--version")
+        {
+            Arity = ArgumentArity.Zero,
+        };
 
-        public static readonly CliOption<bool> InfoOption = new("--info");
+        public static readonly CliOption<bool> InfoOption = new("--info")
+        {
+            Arity = ArgumentArity.Zero,
+        };
 
-        public static readonly CliOption<bool> ListSdksOption = new("--list-sdks");
+        public static readonly CliOption<bool> ListSdksOption = new("--list-sdks")
+        {
+            Arity = ArgumentArity.Zero,
+        };
 
-        public static readonly CliOption<bool> ListRuntimesOption = new("--list-runtimes");
+        public static readonly CliOption<bool> ListRuntimesOption = new("--list-runtimes")
+        {
+            Arity = ArgumentArity.Zero,
+        };
```
Really these look like options but are more like commands - I'd like to model them that way in the future but previous PRs have broken and I haven't gone back to look at it. In any case, these aren't every actually invoked by us (they are handled by the runtime), they just exist here to correctly model our CLI for help/argument parsing purposes.

baronfel commented on src/System.CommandLine.StaticCompletions/CompletionsCommand.cs:
```diff
@@ -0,0 +1,103 @@
+    { }
+
+    private CompletionsCommand(Dictionary<string, IShellProvider> shellMap)
+        : base("completions", "Commands for generating and registering completions for supported shells")
```
TODO: need localizations for the strings introduced in this command..

Evangelink commented:
@baronfel I'm guessing this can only work when the options are statically registered and so it won't work for our new dotnet test experience, right?

baronfel commented:
Depending on the shell we would support 'dynamic completions' - those that require calling into the System.CommandLine completion mechanisms to get values like NuGet packages or TFMs in a project file. Depending on how your system is set up this might be a viable pathway for you. We could chat about it and try to figure out what could work.

baronfel commented on test/System.CommandLine.StaticCompletions.Tests/System.CommandLine.StaticCompletions.Tests.csproj:
```diff
@@ -0,0 +1,28 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- None of these deps should be hard-coded, but something's wonky unless I opt into the whole massive Microsoft.NET.TestFramework mess.
+         I don't want that. This should be able to stand alone. -->
+    <PackageReference Include="System.CommandLine" Version="2.0.0-beta4.24528.1"/>
+    <PackageReference Include="FluentAssertions" Version="6.12.0" />
+    <PackageReference Include="XUnit" Version="2.9.0" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.8.2" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.13.0-preview-24610-01" />
+    <PackageReference Include="Microsoft.TestPlatform" Version="17.13.0-preview-24610-01" />
+    <PackageReference Include="Verify.Xunit" Version="25.0.2" />
+    <PackageReference Include="Verify.DiffPlex" Version="3.0.0" />
```
Note for the SDK team: I'm trying on purpose to keep this test project as slim as possible, but that has two side-effects that I'm having trouble reasoning about:

* the package versions aren't coming through in the way I'd expect, requiring this duplication
* I suspect Helix is mad due to some missing convention

I'm really hoping to not have to take a dependency on the rather large SDK test framework project because that massively expands cycle times.

Forgind replied:
Does central package versioning not get us the versions? Or rather, why not?

baronfel replied:
I can try again and dig into the binlog, but the CPM setup was not working for this project and I didn't want to get blocked on the actual feature work (making tests, etc). Will take another pass at this.

baronfel commented:
The new tests seems to be failing due to some permissions issues - I notice that our version of Verify is quite old. I'm going to trigger package ingestion for that family of packages and then see if bumping those fixes the problem.

baronfel committed:
Fix two logical test failures

baronfel committed:
Remove unnecessary arity and hidden flags from arguments where that materially impacted the end user experience

baronfel committed:
Update Verify packages to hopefully get around some initialization side effects in CI

baronfel committed:
Try to work around Verify directory usage in CI

baronfel committed:
Minor rename to align type name with file name

baronfel committed:
Add dotnet CLI generate script approval tests to the CLI-level tests

baronfel committed:
Try to be even more explicit

am11 commented on src/System.CommandLine.StaticCompletions/shells/BashShellProvider.cs:
```diff
@@ -0,0 +1,222 @@
+        var initialFunctionName = command.FunctionName().MakeSafeFunctionName();
+        return
+            $"""
+            #! /bin/bash
```
A nit: the proper syntax is without the space and this one is recommended to make scripts portable across the distros:

```suggestion
            #!/usr/bin/env bash
```

baronfel replied:
Thanks - I'll apply that locally because it'll also cause changes in the snapshots that I'll need to cover at the same time. Keep the correctness feedback coming - over time the generator can get incrementally better!

Forgind replied:
I have a little hesitation with respect to the security of `#!/usr/bin/env bash`. I've normally used `#!/usr/bin/bash` in the past, and that pinpoints the exact executable to use rather than grabbing the one specified in $PATH, which is more far easier to tamper with.

am11 replied:
/usr/bin/bash is not guaranteed on all platforms and it is recommended to use environment. This is how we are doing it in dotnet repos for several years. Compromised environment is not going to take us far.

No need to delete #! either, it helps testing the script in isolation. I always put it in interpreter as well as include scripts. Does not hurt. CRLF issue is also not limited to the interpreter line.

baronfel committed:
Apply bash env change from @am11 and update snapshots

zivkan commented on src/System.CommandLine.StaticCompletions/shells/BashShellProvider.cs:
```diff
@@ -0,0 +1,222 @@
+        var initialFunctionName = command.FunctionName().MakeSafeFunctionName();
+        return
+            $"""
+            #! /bin/bash
```
The "shebang" needs to be on the first line of the file, not the second. I validated with a script that does `echo "$BASH_VERSION"` and when there was an empty line before `#!/usr/bin/bash`, then running the script from a shell that does not define `BASH_VERSION`, then the output was empty. Removing the empty line from the script made it work.

baronfel replied:
@KalleOlaviNiemitalo is correct - you can see the generated completions in the [snapshot](https://github.com/baronfel/sdk/blob/427f24df7f89933f0c7257f4ad377f9e26eeea1d/test/dotnet-completions.Tests/snapshots/bash/DotnetCliSnapshotTests.VerifyCompletions.verified.sh#L19) for bash, and it looks well-formed in this regard.

zivkan replied:
How did I not know about [this](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/raw-string)?  I'm definitely going to be using it a lot going forward!  It'll be amazing for a bunch of NuGet's test code.

baronfel commented:
@KalleOlaviNiemitalo that's an interesting question. I'm leaning towards no due to ease for this PR. No other competing CLI framework I've seen attempts to support that, though I do see the use case.

baronfel commented:
In addition, that wouldn't be something that source-build could easily support, for example. A user that requires signed scripts could always generate and sign the script themselves.

baronfel commented:
@KalleOlaviNiemitalo said:
> How does this handle CRLF vs. LF line endings in generated scripts?

I anticipated this an instead chose to [always emit newlines only](https://github.com/baronfel/sdk/blob/427f24df7f89933f0c7257f4ad377f9e26eeea1d/src/System.CommandLine.StaticCompletions/shells/BashShellProvider.cs#L46-L47) in the script generation. Each script generator uses this configuration, as it's well-understood on all platforms and keeps the testing more similar so that we don't have to script for OS-specific newlines.

baronfel committed:
try to fix errant path computation in tests

slang25 commented on src/System.CommandLine.StaticCompletions/shells/FishShellProvider.cs:
```diff
@@ -0,0 +1,21 @@
+        """;
+
+    public string GenerateCompletions(System.CommandLine.CliCommand command) => _dynamicCompletionScript;
+}
```
Will this interfere with the baked in completions provided here:
https://github.com/fish-shell/fish-shell/blob/master/share/completions/dotnet.fish

I'm keen to have descriptions on the completions, but I get that it's a bit of a stretch.

baronfel replied:
Thanks for the pointer to this, I didn't know it existed. They are 5 years old however ;) I'm not actually sure how those are used by a fish user (are they just...ambient? or is there a user gesture required to enable them?), but in any case the completions in this PR would require a user to run the `dotnet completions generate fish` command and use that output in their fish configuration, so nothing automatic would happen. 

Separately, if someone implemented the IShellProvider for fish then users would have always-correct, up-to-date completions ready to go! So in short, this PR would make slower but more up to date completions available for fish users, but the groundwork is right there to make fast and up to date completions possible for fish users easily.

baronfel replied:
It looks like if a user [opts into our completions](https://fishshell.com/docs/current/completions.html#where-to-put-completions) they will supersede the built-in definitions:

> By default, Fish searches the following for completions, using the first available file that it finds:
>    * A directory for end-users to keep their own completions, usually ~/.config/fish/completions (controlled by the XDG_CONFIG_HOME environment variable);
>    * A directory for systems administrators to install completions for all users on the system, usually /etc/fish/completions;
>    * A user-specified directory for third-party vendor completions, usually ~/.local/share/fish/vendor_completions.d (controlled by the XDG_DATA_HOME environment variable);
>    * A directory for third-party software vendors to ship their own completions for their software, usually /usr/share/fish/vendor_completions.d;
>    * The completions shipped with fish, usually installed in /usr/share/fish/completions; and
>    * Completions automatically generated from the operating system’s manual, usually stored in ~/.local/share/fish/generated_completions.

baronfel committed:
make the completions command re-nameable

SimonCropp commented:
try

```
public static class VerifyExtensions
{
    public static async Task Verify(CliCommand command, IShellProvider provider, [CallerFilePath] string sourceFile = "")
    {
        var completions = provider.GenerateCompletions(command);
        var settings = new VerifySettings();
        settings.UseDirectory(Path.Combine("snapshots", provider.ArgumentName));
        await Verifier.Verify(target: completions, extension: provider.Extension, settings: settings, sourceFile: sourceFile);
    }
}
```

SimonCropp commented on test/System.CommandLine.StaticCompletions.Tests/VerifyConfiguration.cs:
```diff
@@ -0,0 +1,19 @@
+        VerifyDiffPlex.Initialize(VerifyTests.DiffPlex.OutputType.Compact);
+        // prevent CI systems from trying to make a snapshot directory, maybe?
+        DerivePathInfo((_sourceFile, projectDir, type, method) => new(directory: Path.Combine(projectDir, "snapshots"), typeName: type.Name, methodName: method.Name));
+        EmptyFiles.FileExtensions.AddTextExtension("ps1");
```
hmm. that should be included by default in empty files. i will fix that so this line can be removed in the next update of Verify

baronfel committed:
update verify extensions to forward sourceFile paths

baronfel commented on test/System.CommandLine.StaticCompletions.Tests/VerifyConfiguration.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        VerifyDiffPlex.Initialize(VerifyTests.DiffPlex.OutputType.Compact);
+        // prevent CI systems from trying to make a snapshot directory, maybe?
+        DerivePathInfo((_sourceFile, projectDir, type, method) => new(directory: Path.Combine(projectDir, "snapshots"), typeName: type.Name, methodName: method.Name));
```
@SimonCropp so what I'm trying to do here is tell Verify to use the local `snapshot` directory as the 'base', but then in the [`VerifyExtensions.Verify(CliCommand command, IShellProvider provider)`](https://github.com/dotnet/sdk/pull/42416/files#diff-8ef54fce0890ff3b9b213746e0f5aa9634d7a7f0de3d814a6c1f40a602639fb4R15) method I found that I needed to include the full path to use in `settings.UseDirectory`, so I ended up re-specifying `snapshots` there as well. Is this a correct pattern? I'm not entirely clear on the responsibilities of `DerivePathInfo` vs `settings.UseDirectory`.

baronfel replied:
Why this matters (I think) is that it's impacting the behavior of Verify at test-time in CI - right now it's attempting to access protected directories, and I'm not sure that it _needs_ to if I can correctly configure the snapshot directory. The current error looks like this:

```
System.UnauthorizedAccessException : Access to the path '/Users/runner' is denied.
---- System.IO.IOException : Permission denied
   at System.IO.FileSystem.CreateParentsAndDirectory(String fullPath, UnixFileMode unixCreateMode)
   at System.IO.FileSystem.CreateDirectory(String fullPath, UnixFileMode unixCreateMode)
   at System.IO.Directory.CreateDirectory(String path)
   at IoHelpers.CreateDirectory(String directory) in /_/src/Verify/IoHelpers.cs:line 94
   at VerifyTests.InnerVerifier.ResolveDirectory(String sourceFile, VerifySettings settings, PathInfo pathInfo) in /_/src/Verify/Verifier/InnerVerifier.cs:line 356
   at VerifyTests.InnerVerifier..ctor(String sourceFile, VerifySettings settings, String typeName, String methodName, IReadOnlyList`1 methodParameters, PathInfo pathInfo) in /_/src/Verify/Verifier/InnerVerifier.cs:line 55
   at VerifyXunit.Verifier.BuildVerifier(VerifySettings settings, String sourceFile, Boolean useUniqueDirectory) in /_/src/Verify.Xunit/Verifier.cs:line 18
   at VerifyXunit.Verifier.<>c__DisplayClass9_0.<<Verify>b__0>d.MoveNext() in /_/src/Verify.Xunit/Verifier.cs:line 69
--- End of stack trace from previous location ---
   at Microsoft.DotNet.Cli.Completions.Tests.DotnetCliSnapshotTests.VerifyCompletions(String shellName) in /Users/runner/work/1/s/test/dotnet-completions.Tests/DotnetCliSnapshotTests.cs:line 22
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
```

baronfel replied:
I'm a bit confused because we already have Verify configured success for some Template Engine tests in this same repo, and I attempted to keep the configuration the same as much as possible - though it's very possible it's still some oversight on my end.

SimonCropp replied:
DerivePathInfo is generally used as a way of applying a new (and global) convention for all test. there is an example of a convention here

> method I found that I needed to include the full path to use in settings.UseDirectory

Full path is allowed. but the default is will do a Path.Combine with the directory that contains the current test https://github.com/VerifyTests/Verify/blob/main/docs/naming.md#directory

> I'm not entirely clear on the responsibilities of DerivePathInfo vs settings.UseDirectory.

The default behaviour for DerivePathInfo  is https://github.com/VerifyTests/Verify/blob/main/docs/naming.md#default-derivepathinfo
Using settings.UseDirectory will overide the directory part of that

> Why this matters (I think) is that it's impacting the behavior of Verify at test-time in CI - right now it's attempting to access protected directories, and I'm not sure that it needs to if I can correctly configure the snapshot directory. The current error looks like this:

i dont understand this. So when a test fails, it will attemt to write the snapshot to disk, and hence create the target dir if it exists. but in CI it should not need to create that directory because it should already exist and have files in it

baronfel replied:
Some clues!

```
The directory (/mnt/vss/_work/1/s/test/System.CommandLine.StaticCompletions.Tests/snapshots/zsh) containing the source file (/mnt/vss/_work/1/s/test/System.CommandLine.StaticCompletions.Tests/ZshShellProviderTests.cs) does not exist.
Verify is going to try to recreate the directory and that won't work in CI.
The closest existing directory is (/mnt). The current directory is (/datadisks/disk1/work/B9F10A62/w/965B0881/e)
```

I added some diagnosis code to my verify extensions, and it seems like the tests are being executed in perhaps an entirely different ephemeral container or job? @MiYanni you know things about stuff in the CI system - does this sound normal, or does it trigger any battle scars you have from our build process?

baronfel committed:
Remove DerivePathInfo in tests

If we have corrected paths from the extension methods, maybe we don't need this override?

baronfel committed:
try to debug directory creations

baronfel committed:
even more diagnostics

SimonCropp commented on test/System.CommandLine.StaticCompletions.Tests/VerifyExtensions.cs:
```diff
@@ -13,11 +13,22 @@ public static async Task Verify(CliCommand command, IShellProvider provider, [Ca
         var sourceFileDir = Path.GetDirectoryName(sourceFile)!;
-        if (!Directory.Exists(sourceFileDir))
+        var snapshotDirectory = new DirectoryInfo(Path.Combine(sourceFileDir, "snapshots", provider.ArgumentName));
+        var closestExistingDirectory = GetClosestExistingDirectory(snapshotDirectory);
```
what is snapshotDirectory resolving to in CI?

baronfel replied:
snapshotDirectory: /mnt/vss/_work/1/s/test/System.CommandLine.StaticCompletions.Tests/snapshots/zsh
currentDirectory: /datadisks/disk1/work/B9F10A62/w/965B0881/e

I think there's something odd about the test execution setup/layout, but I don't know enough about the way the CI setup runs test code to narrow it down. Need to chat with the SDK team some more.

baronfel replied:
@SimonCropp I talked with @MiYanni and the tests are indeed executing in a completely different environment, one where source file paths at available at compile-time are completely irrelevant.

I've done some path-munging to correctly find the snapshots in the test environment now, but the verifer is still trying to create a directory that I believe is present and explicitly passed via `settings.UseDirectory`. Is there a way to prevent any directory creation, or are there other mechanisms I can use to customize/sidestep directory location?

baronfel replied:
Got it! https://github.com/dotnet/sdk/pull/42416/commits/823c94010323f8f9a595353fccfeee086319bd72 was what I needed.

baronfel replied:
The key to the whole thing was getting `DerivePathInfo` overriden correctly, which (due to our build processes) meant that I needed to:

* copy snapshots to test output directories
* point the DerivePathInfo delegates to the snapshot directory off of the Environment.CurrentDirectory.

baronfel replied:
Just pushed a change that makes all of that behavior conditional on CI, yes. Good call!

baronfel committed:
try another way of finding paths?

baronfel committed:
fix XMLdoc tags to green up the builds

baronfel committed:
re-set DervicePathInfo

baronfel committed:
remove unused code

baronfel committed:
update .NET CLI snapshot tests similarly to standlone tests

baronfel committed:
Updates to snapshots for new dotnet nuget why documentation

baronfel committed:
Unify newline handling in powershell for very long descriptions

baronfel commented:
All remaining test failures are ephemeral containers failures! Remaining tasks:

* [ ] localization for new help descriptions and such
* [ ] find a way to make the tests not use the deployed snapshots when running locally vs running in CI - the fix I've done harms the local-test comparison experience

MiYanni commented on src/System.CommandLine.StaticCompletions/shells/ZshShellProvider.cs:
```diff
@@ -0,0 +1,371 @@
+            }
+        }
+
+        var catch_all_emitted = false;
```
nit: Is this some kind of contextually named variable? I see others below using this underscore naming style tool.

baronfel replied:
You're right the style is different here - I was transliterating a lot of this from clap-complete, and so I mimicked the code structure and naming of that library to help ease the porting. A subsequent style refactoring could happen, but at least during the writing the similarity helped me keep my place.

MiYanni commented on test/System.CommandLine.StaticCompletions.Tests/System.CommandLine.StaticCompletions.Tests.csproj:
```diff
@@ -0,0 +1,33 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- None of these deps should be hard-coded, but something's wonky unless I opt into the whole massive Microsoft.NET.TestFramework mess.
+         I don't want that. This should be able to stand alone. -->
```
I still don't fully understand the TestFramework avoidance situation. But as it stands, everything running through Helix in this repo should use it. And if it doesn't work well, we should fix it.

MiYanni reviewed (approved):
I left some comments to look over. Generally, I'd avoid implementing a feature like this all in one-shot. Might have made the most sense to make the base-level implementation with 1 shell, and then have PRs for the other shells. I focused on the PowerShell shell since I don't have much experience (or any) with the other shells. But approving anyway since I can't piecemeal approve.

I'm still suspicious on why the TestFramework cannot be used as everything running on Helix should use it.

baronfel committed:
Make local builds use nicer Verify diff compares

By not copying the snapshots to build output, and by making Verify
condition the checks for where to look for the snapshots, we can keep
a good dev experience of diffs on in-tree files instead of digging in
build output directories. CI test failures will still be unfortunately messy.

baronfel commented on src/System.CommandLine.StaticCompletions/CompletionsCommand.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+                return argResult.Tokens switch
+                {
+                // shouldn't be required because of the DefaultValueFactory above
+                [] => LookupShellFromEnvironment(shellMap),
+                [var shellToken] => shellMap[shellToken.Value],
+                    _ => throw new InvalidOperationException("Unexpected number of tokens")
```
I don't know how to solve this one - the auto-formatter keeps/prefers this formatting so anything done manually here to correct it is futile.

SimonCropp commented on test/System.CommandLine.StaticCompletions.Tests/VerifyConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        VerifyDiffPlex.Initialize(VerifyTests.DiffPlex.OutputType.Compact);
+
+        if (IsCI())
```
just an FYI. u could use BuildServerDetector.Detected here https://github.com/VerifyTests/DiffEngine?tab=readme-ov-file#buildserverdetector

baronfel replied:
Excellent pointer - added it to my tests' code.

baronfel committed:
Add localizations for the new strings in the completion and generation commands

baronfel committed:
Remove CI-detection codepaths in favor of DiffEngine's detection

baronfel committed:
Better help descriptions and guidance for the powershell script provider

baronfel commented on src/System.CommandLine.StaticCompletions/shells/PowershellShellProvider.cs:
```diff
@@ -0,0 +1,244 @@
+
+public class PowershellShellProvider : IShellProvider
+{
+    public static string PowerShell => "pwsh";
```
Good call - I've added a 'HelpDescription' member to the IShellProvider interface so that we can show this kind of detail to users. Now they should see it if they call `--help` on the various commands.

baronfel committed:
Generate better completions for the shell arguments

baronfel commented on src/System.CommandLine.StaticCompletions/HelpGenerationExtensions.cs:
```diff
@@ -0,0 +1,121 @@
+    public static string[] Names(this CliOption option)
+    {
+        var (primary, aliases) = PrimaryNameAndAliases(option);
+        return aliases is null ? [primary] : [primary, .. aliases];
```
Similarly to other formatting notes, the formatter keeps inserting this space. I just do what it does? Maybe there's an editorconfig know or something.

baronfel committed:
Update the PowershellShellProvider to use @MiYanni's code and update baselines

baronfel committed:
Try this detection instead

baronfel committed:
Update completions for underlying changes from main

baronfel committed:
Get some test-level output logging about runtime state

baronfel committed:
dump logs to see why my checks aren't working

baronfel committed:
Yield to the borg and onboard to the helix test infra

baronfel committed:
try to detect helix

baronfel committed:
always copy snapshot contents to output directory

baronfel committed:
try to make the scripts not end up in packages?

baronfel committed:
Update assertions due to work merged into main

MiYanni commented on src/System.CommandLine.StaticCompletions/shells/PowershellShellProvider.cs:
```diff
@@ -0,0 +1,248 @@
+        if (argument.IsDynamic())
+        {
+            // if the argument is a not-static-friendly argument, we need to call into the app for completions
+            // TODO: not yet supported for powershell
```
Hmm. I just looked and I thought this feature existed: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_argument_completion?view=powershell-7.5#dynamic-validateset-values-using-classes

baronfel replied:
Nice call out - we could maybe version-probe here? I'd be willing to leave that as an enhancement though

MiYanni replied:
Yeah, I'd just make a new issue to add this feature later.

MiYanni commented on src/System.CommandLine.StaticCompletions/shells/ZshShellProvider.cs:
```diff
@@ -0,0 +1,377 @@
+        using var textWriter = new StringWriter { NewLine = "\n" };
+        using var writer = new IndentedTextWriter(textWriter);
+        string[] pathToCurrentCommand = [command.Name];
+
+
```
nit

baronfel committed:
Update src/System.CommandLine.StaticCompletions/Resources/Strings.resx

Co-authored-by: Michael Yanni <MiYanni@microsoft.com>

baronfel committed:
add missing translations

baronfel committed:
Add some diagnostics for the standalone completions tests

baronfel committed:
fix casing on ps1 script after string change

baronfel committed:
try this way of detecting CI stuff

baronfel committed:
try more helix detection

baronfel commented:
This should be documented for 10 preview 3.
